pixels-turbo/pixels-worker-vhive/src/main/java/io/pixelsdb/pixels/worker/vhive/PixelsReaderStreamImpl.java

```diff
@@ -0,0 +1,528 @@
+                            int hash = Integer.parseInt(req.headers().get("X-Partition-Id"));
+                            if (hash < 0 || hash >= numHashes) {
+                                logger.warn("Client sent invalid hash value: " + hash);
+                                FullHttpResponse response = new DefaultFullHttpResponse(req.protocolVersion(), BAD_REQUEST);
+                                sendResponseAndClose(ctx, req, response);
+                                return;
+                            }
+                            byteBufBlockingMap.put(hash, byteBuf);
+                            if (numHashesReceived.accumulateAndGet(1, Integer::sum) == numHashes) {
+                                // The reader has read all the hashes, so we can put an artificial empty ByteBuf into the queue to signal the end of the stream.
+                                byteBufBlockingMap.put(numHashes, Unpooled.buffer(0).retain());
+                            }
+                        }
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }  // todo: I ignored this and many other exceptions because the method signature is inherited and I cannot throw them out
+                }
+
+                FullHttpResponse response = new DefaultFullHttpResponse(req.protocolVersion(), HttpResponseStatus.OK);
+                sendResponseAndClose(ctx, req, response);
+            }
+
+            private void sendResponseAndClose(ChannelHandlerContext ctx, FullHttpRequest req, FullHttpResponse response) {
+                ChannelFuture f = ctx.writeAndFlush(response);
+                f.addListener(future -> {
+                    if (!future.isSuccess()) {
+                        logger.error("Failed to write response: " + future.cause());
+                        ctx.channel().close();
+                    }
+                });
+                f.addListener(ChannelFutureListener.CLOSE);  // shut down the connection with the current hash
+                if (Objects.equals(req.headers().get(CONNECTION), CLOSE.toString()) || (partitioned && numHashesReceived.get() == numHashes)) {
+                    f.addListener(future -> {
+                        // shutdown the server
+                        ctx.channel().parent().close().addListener(ChannelFutureListener.CLOSE);
+                        // removes schema port to avoid port conflict or misuse
+                        // if (httpPort < 50100)
+                        //     StreamWorkerCommon.delPort(httpPort);
+                    });
+                }
+            }
+        });
+        this.httpServerFuture = CompletableFuture.runAsync(() -> {
+            try {
+                this.httpServer.serve(httpPort);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }, executorService);
+    }
+
+    public static class BlockingMap<K, V> {
```

Comment:
Put this BlockingMap into pixels-core/src/main/java/io/pixelsdb/pixels/core/utils/BlockingMap.java.